pgrx/src/datum/from.rs

```diff
@@ -412,7 +412,22 @@ impl FromDatum for String {
         is_null: bool,
         typoid: pg_sys::Oid,
     ) -> Option<String> {
-        FromDatum::from_polymorphic_datum(datum, is_null, typoid).map(|s: &str| s.to_owned())
+        if varlena::varatt_is_1b_e(datum.cast_mut_ptr::<pg_sys::varlena>())
+            || varlena::varatt_is_4b_c(datum.cast_mut_ptr::<pg_sys::varlena>())
+        {
+            let varlena = pg_sys::pg_detoast_datum_packed(datum.cast_mut_ptr());
+            let converted_varlena = convert_varlena_to_str_memoized(varlena);
+            let ret_string = converted_varlena.to_owned();
+
+            // If the datum is EXTERNAL or COMPRESSED, then detoasting creates a pfree-able chunk
+            // that needs to be freed. We can free it because `to_owned` above creates a Rust copy
+            // of the string.
+            pg_sys::pfree(varlena.cast());
```

Comment:
I'm not feeling at my best right now so perhaps doing code review is a bit ambitious, but I had half-forgotten we were nested inside this `if` by the point of this pfree, when I started to double-check the correctness of this pfree. I had to look back up. Not all code can be so "tidy" but I prefer it when simple implementations with high correctness concerns are in fact that.

So for legibility reasons, I'd prefer if this implementation proceeds *unconditionally* until the pfree, and only *then* checks a condition and then does the pfree based on that condition. Usually Postgres implements this as a pointer-equality check with a comment about how detoasting can duplicate, which also means it doesn't need to rely much on the implementation details of `pg_detoast_datum_packed`, so Postgres can add more conditionals or whatever, but I don't really care.